Fix #2472 Add TFO support to socket BIO and s_client/s_server

tmshort opened the pull request:
Supports Linux, MacOS and FreeBSD

<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING

Other than that, provide a description above this comment if there isn't one already

If this fixes a github issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [X] documentation is added or updated
- [x] tests are added or updated

tmshort commented:
A WIP only because only Linux. FreeBSD and macOS are supported.
This could be a starting point to provide support.
Testing is a bit hard due to platform dependencies.

tmshort commented:
Fixes #2472

levitte commented:
Reading through this change, it looks to me like the availability of TFO can be detected in compile time.  That makes me wonder if a Configure option is needed at all.

tmshort commented:
> Reading through this change, it looks to me like the availability of TFO can be detected in compile time. That makes me wonder if a Configure option is needed at all.

What I have found is that while it can be detected at compile time, it doesn't mean that it is guaranteed to work: FreeBSD, even though it has the correct definitions, doesn't compile it on by default.

In addition, it may be supported under Windows, but I am not a Windows developer!

tmshort commented:
(Could always have it enabled by default, and disabled during configuration?)

mattcaswell commented on apps/s_socket.c:
```diff
@@ -145,9 +146,16 @@ int init_client(int *sock, const char *host, const char *port,
+        if (tfo)
+            options |= BIO_SOCK_TFO;
+#else
+        (void)tfo;
```
This seems unnecessary?

mattcaswell commented on apps/s_socket.c:
```diff
@@ -239,6 +248,12 @@ int do_server(int *accept_sock, const char *host, const char *port,
+    if (tfo)
+        sock_options |= BIO_SOCK_TFO;
+#else
+    (void)tfo;
```
unnecessary?

tmshort replied:
To avoid some compilers complaining of an unused argument when `OPENSSL_NO_TFO` is defined.

mattcaswell replied:
Yeah - I get why you added it. But `tfo` here is a parameter to the function. We have *many* places in the code where we don't use the argument with no problems - so I'm not sure why this one would be.

tmshort replied:
clang is the usual culprit. But I'm reworking it so it won't matter.

mattcaswell commented on Configure:
```diff
@@ -1650,6 +1652,16 @@ unless ($disabled{ktls}) {
+unless ($disabled{tfo}) {
+    $config{tfo}="";
+    if (!($target =~ m/(^linux|^darwin|^BSD)/)) {
+	disable('not-supported', 'tfo');
```
What happens on older linux/darwin/BSD platforms that don't support TCP Fast Open?

tmshort replied:
I have other changes to handle that... but that being said, having it as a configuration option avoids any possible issues. It sounds as though both you and @levitte are against it being a configuration option... If that's the case, I can certainly change it.

mattcaswell commented on Configure:
```diff
@@ -464,6 +465,7 @@ our %disabled = ( # "what"         => "comment"
                   "ssl-trace"           => "default",
                   "ssl3"                => "default",
                   "ssl3-method"         => "default",
+                  "tfo"                 => "default",
```
I am sceptical about having this as a config time option. But if we do have it then it should be added to INSTALL.

mattcaswell reviewed (dismissed):
We should have a test for this

tmshort commented:
> We should have a test for this

Agreed. Suggestions welcome. It's very platform-specific. There are very few ways to detect if TFO was actually used. Linux has global counters. FreeBSD has the TFO #defines, but doesn't enable TFO enabled in the kernel by default.

kaduk commented:
[this will need a rebase past the great local-header rename of #9333]

mspncp commented:
> [this will need a rebase past the great local-header rename of #9333]

@tmshort pull request https://github.com/openssl/openssl/pull/9333#issuecomment-536105158  has an instruction how to rebase over the reorganization of the header files. If that doesn't help, just drop me a note, I'll try to assist with the rebase.

openssl-machine commented:
This PR is waiting for the creator to make requested changes but it has not been updated for 77 days.  If you have made changes or commented to the reviewer please make sure you re-request a review (see icon in the 'reviewers' section).

openssl-machine commented:
This PR is waiting for the creator to make requested changes but it has not been updated for 108 days.  If you have made changes or commented to the reviewer please make sure you re-request a review (see icon in the 'reviewers' section).

openssl-machine commented:
This PR is waiting for the creator to make requested changes but it has not been updated for 139 days.  If you have made changes or commented to the reviewer please make sure you re-request a review (see icon in the 'reviewers' section).

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

t8m commented on apps/s_client.c:
```diff
@@ -2196,6 +2204,17 @@ int s_client_main(int argc, char **argv)
 #endif /* OPENSSL_NO_DTLS */
         sbio = BIO_new_socket(sock, BIO_NOCLOSE);
 
+    if (tfo) {
+        BIO_ADDRINFO *bai;
+
+        if (!BIO_lookup_ex(host, port, BIO_LOOKUP_CLIENT, socket_family,
+                           socket_type, protocol, &bai)) {
+            ERR_print_errors(bio_err);
+            goto shut;
+        }
+        (void)BIO_set_conn_address(sbio, BIO_ADDRINFO_address(bai));
+        BIO_ADDRINFO_free(bai);
+    }
```
The duplicate BIO_lookup_ex call here is inefficient and also not fully correct, because there might be multiple addresses returned and you do not know here which one was actually used by the socket. Could you please refactor init_client to return the used address? Or handle it somehow else?

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

openssl-machine commented:
This PR is waiting for the creator to make requested changes but it has not been updated for 30 days.  If you have made changes or commented to the reviewer please make sure you re-request a review (see icon in the 'reviewers' section).

openssl-machine commented:
This PR is waiting for the creator to make requested changes but it has not been updated for 61 days.  If you have made changes or commented to the reviewer please make sure you re-request a review (see icon in the 'reviewers' section).

kaduk commented:
Is this actually waiting for a contributor response?  I thought we needed some suggestions on how to reliably test it (and maybe about whether a Configure option is appropriate), but neither of those need to come from the contributor.

tmshort commented:
> Is this actually waiting for a contributor response? I thought we needed some suggestions on how to reliably test it (and maybe about whether a Configure option is appropriate), but neither of those need to come from the contributor.

@kaduk, @t8m had requested a refactoring of `init_client` which has been done in this latest update. However, testing is still an outstanding question.

kaduk commented:
> > Is this actually waiting for a contributor response? I thought we needed some suggestions on how to reliably test it (and maybe about whether a Configure option is appropriate), but neither of those need to come from the contributor.
> 
> @kaduk, @t8m had requested a refactoring of `init_client` which has been done in this latest update. However, testing is still an outstanding question.

In that case I will "dismiss" Matt's review so that the openssl-machine reminders will (hopefully) be more accurate

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 61 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 92 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 123 days ago

richsalz commented:
It would be nice to merge this and close out an issue from 2016!

t8m commented on doc/man3/BIO_ctrl.pod:
```diff
@@ -81,6 +86,12 @@ sending. Otherwise, it returns zero.
+BIO_get_conn_mode() returns the BIO connection mode. BIO_set_conn_mode() sets
+the BIO connection mode.
+
+BIO_set_tfo() disables TCP Fast Open when b<onoff> is 0, and enables TCP Fast
+Open when b<onoff> is nonzero.
```
`B<>` or even better `I<>` but that would require changing the argument emphasis for the other functions in the file to `I<>` as well to avoid mixed style within the manpage.

tmshort replied:
Unfortunately, argument highlighting (either via `B<>` or `I<>`) is inconsistent within the documentation, sometimes within the same file... I can at least make this `B<>`.

t8m commented:
@mattcaswell Would this be still acceptable to merge? IMO it looks fairly isolated so should not impact the existing functionality.

tmshort commented:
The `no-sock` check is relatively new. I will have to look at that.

t8m commented:
The sanitizer failure is not relevant.

t8m commented on crypto/bio/bss_conn.c:
```diff
@@ -484,8 +496,21 @@ static long conn_ctrl(BIO *b, int cmd, long num, void *ptr)
         else
             data->connect_mode &= ~BIO_SOCK_NONBLOCK;
         break;
+    case BIO_C_SET_TFO:
```
I'm not sure this should happily return 1 even if TCP_FASTOPEN is not defined and thus not supported.

tmshort replied:
Yeah, if the platform doesn't support TCP_FASTOPEN, this should be ignored.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

tmshort commented:
`fips-sources.checksums` includes `bio.h.in` and `bioerr.h`

levitte reviewed (changes requested):
Please add a line in NEWS.md and a few lines in CHANGES.md

mattcaswell commented:
OTC: Can be reviewed as a normal PR

@openssl/committers - please review

slontis commented:
Needs tests to be added.

tmshort commented:
> Needs tests to be added.

That's actually a bit of a problem; there's no way to tell if TFO is actually being used without a packet capture. There's no data on the socket itself to indicate that TFO had been used. There are system-wide variables that will tell you if TFO has occurred, but because they are system-wide, you can't rely on them for any particular connection.

So, in my first comment I put:

> Testing is a bit hard due to platform dependencies [Edit: suggestions welcome!].

Because TFO is always enabled when available, the most basic test is "did anything break in the regular tests"?

That being said, there is no explicit use of `BIO_set_tfo()` in a test.

tmshort commented:
There's no current tests that use `BIO_s_socket()`. All the test infrastructure uses `BIO_s_mem()` for client/server (there are exceptions for DTLS).

The TLSProxy might be an option, since `tfo` options were added to `s_server` and `s_client`.

So, doing anything requiring TCP sockets may require a new strategy. I would appreciate any suggestions or feedback here.

mattcaswell commented:
> EDIT: Would the TLSProxy interfere with TFO? 

Yes.

I suggest two types of test. One at the BIO level, which simply creates a socket for listening and then connects to it having set the TFO option and checks that the connection is made successfully (optionally pass some data to confirm the socket works).

A second test which fires up s_server and s_client and uses the `-tfo` option and confirm that they talk to each other. This is trickier since there is no infrastructure to support this - but hopefully isn't too difficult. I don't think it is necessary to confirm that TFO is actually *used*. Only that the "-tfo" option doesn't break anything and s_client and s_server can talk to each other. Optionally you could grep the output looking for the "Listening for TFO" message to confirm that the "-tfo" option did *something*.

mattcaswell commented on apps/lib/s_socket.c:
```diff
@@ -315,6 +330,8 @@ int do_server(int *accept_sock, const char *host, const char *port,
     if (sock_family == AF_INET6)
         sock_options |= BIO_SOCK_V6_ONLY;
+    if (tfo)
+        sock_options |= BIO_SOCK_TFO;
```
Is this intentionally within the `ifdef AF_INET6` guard?

mattcaswell replied:
What happens if this is a UDP socket?

tmshort replied:
IPv6 Looks to have been a rebase error.
It should be a no-op for UDP and SCTP, as this sets a flag on the BIO and behavior changes on the connect (which doesn't happen with UDP).

tmshort replied:
Fixed

mattcaswell commented on apps/s_client.c:
```diff
@@ -2013,10 +2022,14 @@ int s_client_main(int argc, char **argv)
     }
+#ifndef OPENSSL_NO_DTLS
+    if (isdtls)
+        tfo = 0;
```
Should this actually be an error?

tmshort replied:
That's pretty subjective, but I suppose it could be.

tmshort replied:
Fixed

mattcaswell commented on crypto/bio/bio_local.h:
```diff
@@ -134,6 +134,12 @@ struct bio_st {
+        BIO_ADDR peer;
+        int first;
+    } tfo;
+#endif
```
Should we really be putting this in the top level bio_st? Wouldn't this be better placed in a BIO specific structure?

tmshort replied:
Can't use `ptr` and the other option is `ex_data`, both of which are user visible, and `ex_data` use is challenging enough that it's easy for a user to break things. I originally had this data stored in `ptr`, but because it's accessible via `BIO_get_data()` it broke some user apps... they should've used `ex_data`, but that's hard to use correctly. I'm actually a little concerned that the `BIO_F_BUFFER_CTX` structure is stored in `ptr`...

This is a fairly small amount of data, and I don't think it's that burdensome. I would want to add a new pointer (since I can't use the user-visible `ptr`) and it subsequently becomes the difference between `sizeof(void*)` vs `sizeof(struct sockaddr) + sizeof(int)`

tmshort replied:
I realize I'm harping on `ptr`, but it bit us pretty bad!

tmshort replied:
Fixed

mattcaswell commented on include/openssl/bio.h.in:
```diff
@@ -427,6 +430,7 @@ struct bio_dgram_sctp_prinfo {
 #  define BIO_get_conn_port(b)          ((const char *)BIO_ptr_ctrl(b,BIO_C_GET_CONNECT,1))
 #  define BIO_get_conn_address(b)       ((const BIO_ADDR *)BIO_ptr_ctrl(b,BIO_C_GET_CONNECT,2))
 #  define BIO_get_conn_ip_family(b)     BIO_ctrl(b,BIO_C_GET_CONNECT,3,NULL)
+#  define BIO_get_conn_mode(b,n)        BIO_ctrl(b,BIO_C_GET_CONNECT,4,NULL)
```
Not a fan of the magic numbers

tmshort replied:
Look at all the other code around it... :/

mattcaswell replied:
Yeah ok. I realised you were just following the pattern of the existing code. It's still horrible though - but another PR.

mattcaswell commented on crypto/bio/bss_sock.c:
```diff
@@ -137,6 +138,15 @@ static int sock_write(BIO *b, const char *in, int inl)
+# endif
+# if defined(OSSL_TFO_SENDTO)
+    if (b->tfo.first) {
+        int peerlen = BIO_ADDR_sockaddr_size(&b->tfo.peer);
```
perhaps `socklen_t` instead of `int`?

mattcaswell commented on crypto/bio/bss_conn.c:
```diff
@@ -425,6 +435,8 @@ static long conn_ctrl(BIO *b, int cmd, long num, void *ptr)
                     ret = -1;
                     break;
                 }
+            } else if (num == 4) {
```
magic numbers :-(

tmshort replied:
This follows the rest of the OpenSSL code in this function. It's a bit out of scope to get rid of magic numbers in this PR (IMO).

mattcaswell commented on crypto/bio/bss_sock.c:
```diff
@@ -205,6 +217,25 @@ static long sock_ctrl(BIO *b, int cmd, long num, void *ptr)
         ret = (b->flags & BIO_FLAGS_IN_EOF) != 0;
         break;
+    case BIO_C_GET_CONNECT:
+        if (ptr != NULL && num == 2) {
```
magic numbers

mattcaswell commented on crypto/bio/bss_sock.c:
```diff
@@ -205,6 +217,25 @@ static long sock_ctrl(BIO *b, int cmd, long num, void *ptr)
+        }
+        break;
+    case BIO_C_SET_CONNECT:
+        if (ptr != NULL && num == 2) {
```
magic numbers

tmshort replied:
Out of scope for this PR... it would require significantly larger changes...

mattcaswell commented on doc/man3/BIO_ctrl.pod:
```diff
@@ -141,8 +165,8 @@ the case of BIO_seek() on a file BIO for a successful operation.
-The BIO_get_ktls_send() and BIO_get_ktls_recv() functions were added in
-OpenSSL 3.0.
+The BIO_get_ktls_send(), BIO_get_ktls_recv() and BIO_get_conn_mode()
+functions were added in OpenSSL 3.0.
```
3.1 now!

tmshort replied:
Fixed

mattcaswell commented on test/bio_socket_test.c:
```diff
@@ -0,0 +1,167 @@
+#include <arpa/inet.h>
+#include <fcntl.h>
+#include <openssl/bio.h>
+#include <netinet/tcp.h>
```
Are all these available on all platforms?

tmshort replied:
I pushed this so I can use the checks to validate other platforms.

tmshort replied:
And... certainly not on windows...

tmshort replied:
Fixed

mattcaswell commented on test/bio_socket_test.c:
```diff
@@ -0,0 +1,167 @@
+        goto err;
+
+    /* ACCEPT SOCKET */
+    afd = socket(AF_INET, SOCK_STREAM, 0);
```
Why not BIO_socket?

tmshort replied:
WIP; I based this on other OpenSSL code that used vanilla socket... I'm slowly moving to use `BIO_` calls as I refine the test.

tmshort replied:
Fixed

mattcaswell commented on test/bio_socket_test.c:
```diff
@@ -0,0 +1,167 @@
+        goto err;
+
+    /* CLIENT SOCKET */
+    cfd = socket(AF_INET, SOCK_STREAM, 0);
```
BIO_socket()?

mattcaswell commented on test/bio_socket_test.c:
```diff
@@ -0,0 +1,167 @@
+     */
+    while (sfd == -1 || !cfd_connected) {
+        /* ACCEPT */
+        sfd = accept(afd, NULL, 0);
```
BIO_accept?

tmshort replied:
Resolved

tmshort commented:
@mattcaswell - I'm in the middle of getting appropriate tests working, so `test/bio_socket_test.c` is definitely a work in progress! :D

tmshort commented:
Looks as though there are *strange* problems building on Windows and there's something weird about the Mac networking... so 90% there. It'll take some time for me to get past that 10% hurdle (have to switch machines).

beldmit commented on apps/s_client.c:
```diff
@@ -456,6 +456,7 @@ typedef enum OPTION_choice {
     OPT_USE_SRTP, OPT_KEYMATEXPORT, OPT_KEYMATEXPORTLEN, OPT_PROTOHOST,
     OPT_MAXFRAGLEN, OPT_MAX_SEND_FRAG, OPT_SPLIT_SEND_FRAG, OPT_MAX_PIPELINES,
     OPT_READ_BUF, OPT_KEYLOG_FILE, OPT_EARLY_DATA, OPT_REQCAFILE,
+    OPT_TFO,
```
#ifdef here?

tmshort replied:
Not really necessary... the actual option definition (`-tfo`) is removed if TCP_FASTOPEN is not defined, so it can't be used if the option is not available.
If this is `#ifdef`'d out, then any reference also has to be `#ifdef`d out, making the code harder to read/follow, IMO.

FdaSilvaYY commented on crypto/bio/bio_tfo.h:
```diff
@@ -0,0 +1,108 @@
+ */
+
+/*
+ * Contains defintions for simplifying the use of TCP Fast Open
```
typo: defintions

FdaSilvaYY commented on doc/man1/openssl-s_server.pod.in:
```diff
@@ -811,6 +812,11 @@ has been negotiated, and early data is enabled on the server. A full handshake
 
+=item B<-tfo>
+
+Enable acceptance of TCP fast open (RFC7413) connections.
```
Typo ? : TCP Fast Open

FdaSilvaYY commented on doc/man3/BIO_s_accept.pod:
```diff
@@ -107,7 +107,9 @@ B<BIO_BIND_REUSEADDR> is set then other sockets can bind to the
 and the port is not in use then a second attempt is made
-using B<BIO_BIND_REUSEADDR>.
+using B<BIO_BIND_REUSEADDR>. If B<BIO_SOCK_TFO> is set, then
+the socket will be configured to accept TCP fast open
```
Fast Open ?

tmshort commented:
Had to add configuration for TFO, and it's off by default. Windows, VMS and other BSDs make it too complicated to have sane tests and enabled by default.

mattcaswell commented:
@tmshort...is this ready for further review or are you still "tweaking"?

tmshort commented:
Added `s_client`/`s_server` test.

mattcaswell commented on NEWS.md:
```diff
@@ -23,6 +23,8 @@ OpenSSL 3.1
   * Subject or issuer names in X.509 objects are now displayed as UTF-8 strings
     by default.
+  * TCP Fast Open support is available on Windows, Linux, macOS, and FreeBSD
+    where enabled and supported.
```
INSTALL.md says "Linux, macOS and FreeBSD" and here we say "Windows, Linux, macOS, and FreeBSD"

mattcaswell commented on test/recipes/82-test_tfo_cli.t:
```diff
@@ -0,0 +1,100 @@
+    print STDERR "Server attempting TFO\n" if ! ok($server_good);
+
+    print STDERR "Connection failed\n" if ! ok($connect_good);
+}
```
Looks like these two tests are essentially copies of each other with some minor differences. Probably there are opportunities to reduce code duplication here and have a common function for these two variants of the same thing.

tmshort replied:
The problem with the connectivity test is that there's no guarantee that it will connect in the CI. The tests pass when built locally.

mattcaswell replied:
Hmmm. We open sockets etc with the TLSProxy tests and these reliably pass - so someway or other it should be possible.

tmshort replied:
I think I see it... let me try something...

tmshort replied:
It was non-obvious that `s_client` would accept a `0` port.

mattcaswell reviewed (changes requested):
CI failures look relevant

mattcaswell commented on test/bio_tfo_test.c:
```diff
@@ -0,0 +1,358 @@
+    }
+
+    /* ACCEPT SOCKET */
+    if (!TEST_ptr(abio = BIO_new_accept("127.0.0.1:0"))
```
What will happen if we run this on an IPv6 only host?

tmshort replied:
Not run? HA! Invariably, everything is dual-stack IPv4+IPv6 or just IPv4. I will update.

mattcaswell replied:
The reason I ask is I *think* we ran into some oddball platforms like that with the TLSProxy code. I could be wrong though.

tmshort replied:
It's now IPv4/IPv6 compatible.

mattcaswell commented on test/recipes/82-test_tfo_cli.t:
```diff
@@ -0,0 +1,85 @@
+setup("test_tfo");
+
+plan skip_all => "test_tfo_cli needs sock enabled" if disabled("sock");
+plan skip_all => "test_tfo_cli needs tls enabled" if disabled("tls");
```
It seems to need a TLS version <1.3 enabled

tmshort replied:
Is `tls` not sufficient for this, then? Not sure what you're asking for here?

mattcaswell replied:
"tls" checks if *a* TLS version is enabled. In the event that *only* TLSv1.3 is enabled then the test would not be skipped, and would then fail.

tmshort replied:
Fixed

tmshort commented on CHANGES.md:
```diff
@@ -24,6 +24,11 @@ OpenSSL 3.1
 
 ### Changes between 3.0 and 3.1 [xx XXX xxxx]
 
+ * Add support for TCP Fast Open to Windows, macOS, Linux, FreeBSD where
```
Is `NEWS.md` redundant? It seems to be an abridged version of `CHANGES.md`.

mattcaswell replied:
See:
https://github.com/openssl/technical-policies/pull/18/files#diff-aeb01d3c36ee32e935f6ea6f32833c5228de657d3783beecc3db0d77c6265599R101-R108

tmshort replied:
I made my opinion known on that PR, thanks.

tmshort commented:
For some reason, github is not letting me comment on this:
> This is the way BIOs are supposed to do things. Applications misusing this are broken apps IMO. They shouldn't be messing with BIO internal data.

Those APIs are dangerous; and perhaps a better warning should be given in the documentation. It allows an application (that's not using a custom BIO) to inadvertently break BIO structures.

mattcaswell commented:
Oh....err...except there are some failing CIs...:-(

tmshort commented:
Yup... looking at those.

tmshort commented:
Ok, it looks as though there's a conflict between TFO and KTLS in the use of `BIO->ptr`.
It appears that KTLS is storing a byte value into `BIO->ptr`:
https://github.com/openssl/openssl/blob/5719dd461fc2cc5d5d29fc3d7e9a6deca3130a7e/crypto/bio/bss_sock.c#L133
and
https://github.com/openssl/openssl/blob/5719dd461fc2cc5d5d29fc3d7e9a6deca3130a7e/crypto/bio/bss_sock.c#L197
Suggestions? I could add a byte to the TFO structure... which means that structure is needed for both TFO and KTLS, and I'd want to look at other places KTLS manipulates `ptr`.

mattcaswell commented:
In general there should be one structure per BIO. So we need a combined structure for bss_sock which stores both TFO and KTLS things.

mattcaswell commented on test/recipes/82-test_tfo_cli.t:
```diff
@@ -0,0 +1,87 @@
+plan skip_all => "test_tfo_cli needs sock enabled" if disabled("sock");
+plan skip_all => "test_tfo_cli needs tls < 1.3 enabled"
+    if disabled("tls1_0") && disabled("tls1_1") && disabled("tls1_2");
+plan skip_all => "test_tfo_cli needs tls enabled" if disabled("tls");
```
It's quite confusing that this is necessary. If `disabled("tls")` is true then this implies that `disabled("tls1_0") && disabled("tls1_1") && disabled("tls1_2")` is also true. So this `if disabled("tls")` clause should be entirely redundant......

.....AH!! except the first skip is incorrect it should read:

`if disabled("tls1") && disabled("tls1_1") && disabled("tls1_2")`

Note that "tls1_0" is the incorrect name - it should be "tls1"

tmshort replied:
Right because it was running when it wasn't supposed to!
One would think that `tls1_0` and `tls1` were aliases...Oh well.. quick patch.

mattcaswell commented:
Ping @openssl/committers. Needs a second review.

beldmit commented on apps/s_client.c:
```diff
@@ -2013,10 +2022,18 @@ int s_client_main(int argc, char **argv)
+        BIO_printf(bio_err, "%s: DTLS does not support the -tfo option\n", prog);
+        goto end;
+    }
+#endif
```
Don't we need smth like this in s_server?

tmshort replied:
Added check in fixup.

tmshort replied:
How the heck is this failing CI now on `#include "e_os.h"` when all I did was add a fixup commit for `s_server`?
I'm not seeing any more recent changes on master that would impact this?

tmshort replied:
Looks like d5f9166bac on master broke it.

tmshort commented:
I had to rebase to get around the moving of `e_os.h`

mattcaswell reviewed (approved):
Reconfirm

kaduk commented on test/bio_tfo_test.c:
```diff
@@ -0,0 +1,398 @@
+        else if (idx == 1)
+            BIO_printf(bio_err, "Ignore: connection accepted before data, possibly no TFO cookie, or TFO may not be enabled\n");
+        else
+            BIO_printf(bio_err, "Warning: connection accepted before data, TFP may not be enabled\n");
```
```suggestion
            BIO_printf(bio_err, "Warning: connection accepted before data, TFO may not be enabled\n");
```

tmshort replied:
Fixed

kaduk commented on test/bio_tfo_test.c:
```diff
@@ -0,0 +1,398 @@
+            || !TEST_long_gt(BIO_set_conn_port(cbio, port), 0)
+            || !TEST_long_gt(BIO_set_nbio(cbio, 1), 0)
+            || !TEST_long_gt(BIO_set_tfo(cbio, client_tfo), 0))
+        goto err;
```
Should either of these first two blocks try to set sockerr?

tmshort replied:
probably

kaduk commented on test/bio_tfo_test.c:
```diff
@@ -0,0 +1,398 @@
+
+    /* SEND DATA: this should establish the actual TFO connection */
+    if (!TEST_true(BIO_write_ex(cbio, SOCKET_DATA, SOCKET_DATA_LEN, &bytes)))
+        goto err;
```
sockerr was last set up at line 126 if we take this error path, which is probably not a useful value anymore.
If we're not going to update it, we should perhaps clear it back to zero closer to when we last set it..

tmshort replied:
It should be updated as the `BIO_write_ex()` could fail.

kaduk commented on test/bio_tfo_test.c:
```diff
@@ -0,0 +1,398 @@
+    /* FINAL ACCEPT: if TFO is enabled, socket should be accepted at *this* point */
+    if (BIO_do_accept(abio) <= 0) {
+        BIO_printf(bio_err, "Error: socket not accepted\n");
+        goto err;
```
we should set sockerr here?

tmshort replied:
probably

kaduk commented on test/bio_tfo_test.c:
```diff
@@ -0,0 +1,398 @@
+            || !TEST_true(BIO_read_ex(sbio, read_buffer, sizeof(read_buffer), &bytes))
+            || !TEST_size_t_eq(bytes, SOCKET_DATA_LEN)
+            || !TEST_strn_eq(read_buffer, SOCKET_DATA, SOCKET_DATA_LEN))
+        goto err;
```
we could set sockerr here as well.

tmshort replied:
fixed

kaduk commented on test/bio_tfo_test.c:
```diff
@@ -0,0 +1,398 @@
+    bytes_read = readsocket(sfd, read_buffer, sizeof(read_buffer));
+    if (!TEST_int_eq(bytes_read, SOCKET_DATA_LEN)
+            || !TEST_strn_eq(read_buffer, SOCKET_DATA, SOCKET_DATA_LEN))
+        goto err;
```
sockerr might be stale here

tmshort replied:
fixed

kaduk commented on test/recipes/82-test_tfo_cli.t:
```diff
@@ -0,0 +1,86 @@
+    print $cin "GET /\r\n";
+
+    waitpid($cpid, 0);
+    kill 'KILL', $spid;
```
Is it better to use `s_sever -naccept 1` than to use `kill 'KILL'`?

tmshort replied:
Yup... that works... and is cleaner. Thanks.

So... many... options...

kaduk commented on test/recipes/82-test_tfo_cli.t:
```diff
@@ -0,0 +1,86 @@
+    my $cpid = open2(my $cout, my $cin, $shlib_wrap, $apps_openssl, @c_cmd);
+
+    # Do the "GET", which will cause the client to finish
+    print $cin "GET /\r\n";
```
```suggestion
    print $cin "GET / HTTP/1.0\r\n\r\n";
```

tmshort replied:
This doesn't really matter. The `-www`/`-WWW` prefix responds to just `GET /`

https://github.com/openssl/openssl/blob/14db620282bea38dc44479e562cf9bb61a716444/apps/s_server.c#L3225

kaduk reviewed (commented):
Much of these are no big deal, but I'll call out the removal of a `#define` from a public header, the question of what the coding-style policy says about macros returning lvalues, and the recurring question of when we should(n't) set `sockerr` in the unit tests.

Full disclosure: I reviewed an internal version of this patchset around when this PR was first created, some two and a half years ago.  That's plenty of time for me to have forgotten anything I might have known...

mattcaswell commented on include/openssl/bioerr.h:
```diff
@@ -44,7 +44,8 @@
 # define BIO_R_NO_HOSTNAME_OR_SERVICE_SPECIFIED           144
 # define BIO_R_NO_PORT_DEFINED                            113
 # define BIO_R_NO_SUCH_FILE                               128
-# define BIO_R_NULL_PARAMETER                             115 /* unused */
```
Oh - good catch. No this isn't ok.

tmshort replied:
I think it was possibly deleted when this was a 3.0 PR. I put it back since it's now 3.1

tmshort commented on crypto/bio/bss_sock.c:
```diff
@@ -194,7 +224,7 @@ static long sock_ctrl(BIO *b, int cmd, long num, void *ptr)
     case BIO_CTRL_SET_KTLS_TX_SEND_CTRL_MSG:
         BIO_set_ktls_ctrl_msg_flag(b);
-        b->ptr = (void *)num;
+        BIO_KTLS_RECORD(b) = (unsigned char)num;
```
See prior comment.

tmshort replied:
fixed

tmshort commented on crypto/bio/bss_sock.c:
```diff
@@ -205,6 +235,25 @@ static long sock_ctrl(BIO *b, int cmd, long num, void *ptr)
+            ret = BIO_ADDR_make(&BIO_TFO_PEER(b),
+                                BIO_ADDR_sockaddr((const BIO_ADDR*)ptr));
+            if (ret)
+                BIO_TFO_FIRST(b) = 1;
```
See prior response.

tmshort replied:
fixed

tmshort commented on crypto/bio/bss_sock.c:
```diff
@@ -157,10 +180,17 @@ static long sock_ctrl(BIO *b, int cmd, long num, void *ptr)
         b->shutdown = (int)num;
         b->init = 1;
+        BIO_TFO_FIRST(b) = 0;
+        memset(&BIO_TFO_PEER(b), 0, sizeof(BIO_TFO_PEER(b)));
```
Nothing technically wrong with it, it keeps the code cleaner and understandable vs. the expansion.

tmshort replied:
fixed

tmshort commented on crypto/bio/bss_sock.c:
```diff
@@ -157,10 +180,17 @@ static long sock_ctrl(BIO *b, int cmd, long num, void *ptr)
     switch (cmd) {
     case BIO_C_SET_FD:
-        sock_free(b);
+        /* minimal sock_free() */
```
Other BIOs don't call their `free` equivalent when handling this control message. They do a minimal free; the problem is that we don't want to free `ptr`.

tmshort replied:
And flags aren't always saved, if `shutdown` is set.

tmshort commented on crypto/bio/bss_sock.c:
```diff
@@ -130,13 +144,22 @@ static int sock_write(BIO *b, const char *in, int inl)
+        ret = sendto(b->num, in, inl, OSSL_TFO_SENDTO,
+                     BIO_ADDR_sockaddr(&BIO_TFO_PEER(b)) peerlen);
+        BIO_TFO_FIRST(b) = 0;
+    } else
```
KTLS did the same thing immediately above this.

tmshort commented on crypto/bio/bss_sock.c:
```diff
@@ -130,13 +144,22 @@ static int sock_write(BIO *b, const char *in, int inl)
+
+        ret = sendto(b->num, in, inl, OSSL_TFO_SENDTO,
+                     BIO_ADDR_sockaddr(&BIO_TFO_PEER(b)) peerlen);
+        BIO_TFO_FIRST(b) = 0;
```
Per my other response, the expansion is quite ugly, or the use of a temporary is required (as in other cases).

tmshort commented on CHANGES.md:
```diff
@@ -24,6 +24,11 @@ OpenSSL 3.1
 
 ### Changes between 3.0 and 3.1 [xx XXX xxxx]
 
+ * Add support for TCP Fast Open to macOS, Linux, FreeBSD where supported
```
Fixed

tmshort commented on apps/lib/s_socket.c:
```diff
@@ -160,14 +167,22 @@ int init_client(int *sock, const char *host, const char *port,
 
+        /* Save the address */
+        if (tfo && ba_ret != NULL)
+            *ba_ret = BIO_ADDR_dup(BIO_ADDRINFO_address(ai));
```
`BIO_ADDRINFO_address()` does a NULL check then returns a value, or NULL. It's not significant enough to require caching.

tmshort commented on doc/man3/BIO_ctrl.pod:
```diff
@@ -144,6 +168,8 @@ the case of BIO_seek() on a file BIO for a successful operation.
 The BIO_get_ktls_send() and BIO_get_ktls_recv() functions were added in
 OpenSSL 3.0.
 
+The BIO_get_conn_mode() function was added in OpenSSL 3.1.
```
fixed.

tmshort commented on crypto/bio/bio_tfo.h:
```diff
@@ -0,0 +1,122 @@
+ * is to go into the /proc filesystem, but I'm not sure it's
+ * worthwhile.
+ *
+ * The OSSL_TFO_xxxxxx_FLAGs can be OR'd together:
+ * 0 = TFO disabled
+ * 1 = client TFO enabled
+ * 2 = server TFO enabled
+ * 3 = server and client TFO enabled
```
It's a bit more complicated than that... but ok.

tmshort commented on crypto/bio/bio_tfo.h:
```diff
@@ -0,0 +1,122 @@
+# endif
+
+# if defined(__FreeBSD__)
+#  define OSSL_TFO_SYSCTL               "net.inet.tcp.fastopen.enabled"
```
Yeah, I'm re-working BSD a little bit, since the support changed at 12.0.0; this will take more time (as I have to now install two versions of FreeBSD somewhere...)

tmshort commented on doc/man3/BIO_s_accept.pod:
```diff
@@ -87,6 +88,9 @@ B<host_port>.
 
+BIO_set_tfo_accept() enables TCP Fast Open on the accept socket
+if B<n> is 1 or disables TCP Fast Open if B<n> is 0 (the default).
+
```
`BIO_set_tfo_accept()` is for `BIO`s. `BIO_SOCK_TFO` is for raw  file descriptors/sockets.

tmshort replied:
^^^ Generally... except in the `BIO_set_bind_mode()`... These are basically the same...

tmshort commented on doc/man3/BIO_ctrl.pod:
```diff
@@ -81,6 +86,12 @@ sending. Otherwise, it returns zero. It also returns negative values for failure
+the BIO connection mode.
+
+BIO_set_tfo() disables TCP Fast Open when B<onoff> is 0, and enables TCP Fast
+Open when B<onoff> is nonzero.
```
Fixed.

tmshort commented:
Updated tests to work with older Linux and FreeBSD versions that support TFO.
Updated documentation
Updated use of `sockerr`
Updated the use of the (renamed) `bss_conn_st` structure

tmshort commented:
ping @mattcaswell reconfirm?

mattcaswell commented on include/internal/bio_tfo.h:
```diff
@@ -0,0 +1,151 @@
+ * determine if TFO is enabled for the server and client respectively.
+ *
+ * OSSL_TFO_SERVER_FLAG = 1 = client TFO enabled
+ * OSSL_TFO_CLIENT_FLAG = 2 = server TFO enabled
```
Errr....OSSL_TFO_CLIENT_FLAG indicates the *server* is enabled???

tmshort replied:
Copying error. Thanks. It get's complicated because on Linux/Mac two bits of a single field are used to indicate enablement, but on FreeBSD, there are separate fields that indicate enablement. The flag values are correct, the comment following is wrong.

tmshort commented:
ping @mattcaswell @kaduk ?

tmshort commented:
ping @levitte @t8m ?

t8m commented on test/bio_tfo_test.c:
```diff
@@ -0,0 +1,418 @@
+    }
+
+    /* macOS needs some time for this to happen, so put in a select */
+    if (!TEST_int_ge(BIO_wait(abio, time(NULL)+2, 0), 0)) {
```
Nit: spaces around `+`

t8m commented on test/bio_tfo_test.c:
```diff
@@ -0,0 +1,418 @@
+    }
+
+    /* macOS needs some time for this to happen, so put in a select */
+    if (!TEST_int_ge(BIO_wait(abio, time(NULL)+2, 0), 0)) {
```
Nit: spaces around `+`

t8m commented on test/bio_tfo_test.c:
```diff
@@ -0,0 +1,418 @@
+    }
+
+    /* macOS needs some time for this to happen, so put in a select */
+    if (!TEST_int_ge(BIO_socket_wait(afd, 1, time(NULL)+2), 0)) {
```
Nit: spaces around `+`

t8m commented on test/bio_tfo_test.c:
```diff
@@ -0,0 +1,418 @@
+    /* SEND DATA: this should establish the actual TFO connection */
+#ifdef OSSL_TFO_SENDTO
+    if (!TEST_int_ge(sendto(cfd, SOCKET_DATA, SOCKET_DATA_LEN, OSSL_TFO_SENDTO,
+                            (struct sockaddr*)&sstorage, slen), 0)) {
```
Nit: space before `*`

tmshort replied:
So, AFAICT, there's no coding specification for casts. The <space> is not strictly necessary. But the reasoning is that by putting the `*` next to the type, it's a bit more obvious _in declarations_ that it's a pointer-type. Of course, this only works well when there's a single declaration per line. And that's why I tend to do that in casts (because of how I do declarations).

t8m replied:
We, throughout the code use space before `*` both in casts and declarations. So thank you for changing this.

t8m commented on crypto/bio/bio_sock2.c:
```diff
@@ -292,6 +357,54 @@ int BIO_listen(int sock, const BIO_ADDR *addr, int options)
+#   endif
+#  endif
+        if (setsockopt(sock, IPPROTO_TCP, OSSL_TFO_SERVER_SOCKOPT,
+                       (void*)&q, sizeof(q)) < 0) {
```
Nit: space before `*`

t8m commented on crypto/bio/bss_sock.c:
```diff
@@ -126,17 +139,30 @@ static int sock_read(BIO *b, char *out, int outl)
 {
     int ret = 0;
+# if !defined(OPENSSL_NO_KTLS) || defined(OSSL_TFO_SENDTO)
+    struct bss_sock_st *data = (struct bss_sock_st*)b->ptr;
```
Nit: space before `*`

t8m commented on crypto/bio/bss_sock.c:
```diff
@@ -126,17 +139,30 @@ static int sock_read(BIO *b, char *out, int outl)
+# endif
+# if defined(OSSL_TFO_SENDTO)
+    if (data->tfo_first) {
+        struct bss_sock_st *data = (struct bss_sock_st*)b->ptr;
```
Nit: space before `*`

t8m commented on crypto/bio/bss_sock.c:
```diff
@@ -151,16 +177,24 @@ static long sock_ctrl(BIO *b, int cmd, long num, void *ptr)
 {
     long ret = 1;
     int *ip;
+    struct bss_sock_st *data = (struct bss_sock_st*)b->ptr;
```
Nit: space before `*`

t8m commented on crypto/bio/bss_sock.c:
```diff
@@ -205,6 +239,25 @@ static long sock_ctrl(BIO *b, int cmd, long num, void *ptr)
+    case BIO_C_SET_CONNECT:
+        if (ptr != NULL && num == 2) {
+            ret = BIO_ADDR_make(&data->tfo_peer,
+                                BIO_ADDR_sockaddr((const BIO_ADDR*)ptr));
```
Nit: space before `*`

t8m commented on test/bio_tfo_test.c:
```diff
@@ -0,0 +1,418 @@
+
+    switch (ai->ai_family) {
+        case AF_INET:
+            port = ((struct sockaddr_in*)ai->ai_addr)->sin_port;
```
Nit: space before `*` (and similarly in all cases below).

t8m commented on test/recipes/82-test_tfo_cli.t:
```diff
@@ -0,0 +1,86 @@
+        }
+    }
+    print STDERR "Port: $port\n";
+
```
Should there be explicit failure from the recipe if we do not get a port, i.e. $port == 0?

tmshort replied:
Possibly.

tmshort replied:
Done

t8m reviewed (changes requested):
Great work! Only some very minor nits left.

tmshort commented on doc/man3/BIO_ADDR.pod:
```diff
@@ -37,7 +39,11 @@ BIO_ADDR_new() creates a new unfilled B<BIO_ADDR>, to be used
 BIO_accept_ex().
 
-BIO_ADDR_free() frees a B<BIO_ADDR> created with BIO_ADDR_new().
+BIO_ADDR_dup() creates a new B<BIO_ADDR>, with a copy of the
```
Done

tmshort commented on doc/man1/openssl-s_server.pod.in:
```diff
@@ -816,6 +817,11 @@ has been negotiated, and early data is enabled on the server. A full handshake
 is forced if a session ticket is used a second or subsequent time. Any early
 data that was sent will be rejected.
 
+=item B<-tfo>
```
Done

tmshort commented on doc/man1/openssl-s_client.pod.in:
```diff
@@ -782,6 +783,10 @@ Enable Kernel TLS for sending and receiving.
 This option was introduced in OpenSSL 3.1.0.
 Kernel TLS is off by default as of OpenSSL 3.1.0.
 
+=item B<-tfo>
```
Done

tmshort commented on doc/man3/BIO_s_accept.pod:
```diff
@@ -23,6 +23,7 @@ BIO_set_bind_mode, BIO_get_bind_mode, BIO_do_accept - accept BIO
  BIO *BIO_new_accept(char *host_port);
 
  long BIO_set_nbio_accept(BIO *b, int n);
+ long BIO_set_tfo_accept(BIO *b, int n);
```
Done

tmshort commented:
@t8m fixed nits.

t8m commented:
@mattcaswell still OK?

mattcaswell reviewed (approved):
Still ok

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but this PR has failing CI tests.  Once the tests pass it will get moved to 'approval: ready to merge' automatically, alternatively please review and set the label manually.

t8m commented:
Aargh another rebase needed. I am sorry for that @tmshort Could you please do it?

t8m reviewed (approved):
Still OK if CI passes.

openssl-machine commented:
This pull request is ready to merge

t8m commented:
Please wait with merging this before the https://github.com/openssl/openssl/pull/17838 is merged

t-j-h commented:
Note added approval on the rebase given I walked through all of this logic when checking #17838

tmshort commented:
Pushed to master.

t8m commented:
There was a typo in the PR number merging this (you've used #8692 instead). But that isn't a big deal.

It also very nicely increased the test coverage, thank you for that.
https://coveralls.io/builds/47264166

tmshort commented:
@t8m Whoops! Sorry about that.

tmshort commented:
Fixes #2472
